src/workerd/io/actor-sqlite.c++

```diff
@@ -157,11 +158,68 @@ void ActorSqlite::onWrite() {
       // rather than after the callback.
       { auto drop = kj::mv(txn); }
 
-      return commitCallback();
+      return commitImpl();
     })));
   }
 }
 
+kj::Promise<void> ActorSqlite::commitImpl() {
+  bool needsAlarmFlush = false;
+  kj::Maybe<kj::Date> newAlarmTime;
+  KJ_SWITCH_ONEOF(currentAlarmTime) {
+    KJ_CASE_ONEOF(_, UnknownAlarmTime) {
+      // Haven't tried to write alarm yet, so don't need to flush.
+    }
+    KJ_CASE_ONEOF(knownAlarmTime, KnownAlarmTime) {
+      if (knownAlarmTime.status == KnownAlarmTime::Status::DIRTY) {
+        knownAlarmTime.status = KnownAlarmTime::Status::FLUSHING;
+        needsAlarmFlush = true;
+        newAlarmTime = knownAlarmTime.time;
+      }
+    }
+    KJ_CASE_ONEOF(deferredDelete, DeferredAlarmDelete) {
+      if (deferredDelete.status == DeferredAlarmDelete::Status::READY) {
+        deferredDelete.status = DeferredAlarmDelete::Status::FLUSHING;
+        needsAlarmFlush = true;
+        newAlarmTime = kj::none;
+      }
+    }
+  }
+
+  // TODO(soon): ActorCache implements a 4x retry of failed flushes... Do we need anything similar
+  // for alarm scheduling?
+
+  // We assume that exceptions thrown during commit will propagate to the caller, such that they
+  // will ensure cancelDeferredAlarmDeletion() is called, if necessary.
+
+  if (needsAlarmFlush) {
+    // TODO(soon): fix sequencing of alarm scheduling vs. commitCallback().
+    co_await hooks.setAlarm(newAlarmTime);
```

Comment:
A couple things missing here:

- We only want to update the alarm time _before_ the commit if the new time is sooner than what was already set. It looks like here you're performing the pre-commit update regardless of which direction the alarm is moving. Of course, this implies that we need to keep track of what the previous value was.
- If, during the call to `setAlarm()`, the alarm value is updated again, and is updated to be sooner than what had been passed to `setAlarm()`, then we need to repeat the `setAlarm()` with the newer value before committing.